lib/states.js

```diff
@@ -28,24 +30,33 @@ Each "state" in the lifetime is a well defined function:
 @constructor
 @param {Array[Object]} hooks for handling states in the task lifecycle.
 */
-function States(hooks) {
-  assert.ok(Array.isArray(hooks), 'hooks is an array');
-  this.hooks = hooks;
-}
-
-States.prototype = {
+export default class States {
+  constructor(hooks) {
+    assert.ok(Array.isArray(hooks), 'hooks is an array');
+    this.hooks = hooks;
+  }
 
   /**
   Invoke all hooks with a particular method.
   @param {Task} task handler.
   */
-  _invoke: function* (method, task) {
-    return yield this.hooks.
-      filter(hasMethod.bind(this, method)).
-      map(function(hook) {
-        return hook[method](task);
-      });
-  },
+  async _invoke(method, task) {
+    let hooks = this.hooks.filter(hasMethod.bind(this, method));
+    let links = await Promise.all(hooks.map((hook) => {
+        return new Promise(async (accept, reject) => {
```

Comment:
The decent two-liner is:
```js
let hooks = this.hooks.filter((hook) => { return hook[method] instanceof Function });
return Promise.all(hooks.map((hook) => { return hook[method](task); }));
```
Maybe one can do the `filter` on the same line, I couldn't make it readable :)